src/mimeparser.rs

```diff
@@ -0,0 +1,1143 @@
+use crate::message::MsgId;
+use crate::param::*;
+use crate::peerstate::Peerstate;
+use crate::securejoin::handle_degrade_event;
+use crate::stock::StockMessage;
+
+#[derive(Debug)]
+pub struct MimeParser<'a> {
+    pub context: &'a Context,
+    pub parts: Vec<Part>,
+    pub header: HashMap<String, String>,
+    pub subject: Option<String>,
+    pub is_send_by_messenger: bool,
+    pub decrypting_failed: bool,
+    pub encrypted: bool,
+    pub signatures: HashSet<String>,
+    pub gossipped_addr: HashSet<String>,
+    pub is_forwarded: bool,
+    pub is_system_message: SystemMessage,
+    pub location_kml: Option<location::Kml>,
+    pub message_kml: Option<location::Kml>,
+    reports: Vec<Report>,
+    mdns_enabled: bool,
+    parsed_protected_headers: bool,
+}
+
+#[derive(Debug, Display, Clone, Copy, PartialEq, Eq, FromPrimitive, ToPrimitive, ToSql, FromSql)]
+#[repr(i32)]
+pub enum SystemMessage {
+    Unknown = 0,
+    GroupNameChanged = 2,
+    GroupImageChanged = 3,
+    MemberAddedToGroup = 4,
+    MemberRemovedFromGroup = 5,
+    AutocryptSetupMessage = 6,
+    SecurejoinMessage = 7,
+    LocationStreamingEnabled = 8,
+    LocationOnly = 9,
+}
+
+impl Default for SystemMessage {
+    fn default() -> Self {
+        SystemMessage::Unknown
+    }
+}
+
+const MIME_AC_SETUP_FILE: &str = "application/autocrypt-setup";
+
+impl<'a> MimeParser<'a> {
+    pub fn from_bytes(context: &'a Context, body: &[u8]) -> Result<Self> {
+        let mail = mailparse::parse_mail(body)?;
+        let mdns_enabled = context.get_config_bool(Config::MdnsEnabled);
+
+        let mut parser = MimeParser {
+            parts: Vec::new(),
+            header: Default::default(),
+            subject: None,
+            is_send_by_messenger: false,
+            decrypting_failed: false,
+            encrypted: false,
+            signatures: Default::default(),
+            gossipped_addr: Default::default(),
+            is_forwarded: false,
+            context,
+            reports: Vec::new(),
+            is_system_message: SystemMessage::Unknown,
+            location_kml: None,
+            message_kml: None,
+            mdns_enabled,
+            parsed_protected_headers: false,
+        };
+
+        let message_time = mail
+            .headers
+            .get_first_value("Date")?
+            .and_then(|v| mailparse::dateparse(&v).ok())
+            .unwrap_or_default();
+
+        parser.hash_header(&mail.headers);
+
+        // Memory location for a possible decrypted message.
+        let mail_raw;
+
+        let mail = match e2ee::try_decrypt(parser.context, &mail, message_time) {
+            Ok((raw, signatures)) => {
+                // Valid autocrypt message, encrypted
+                parser.encrypted = raw.is_some();
+                parser.signatures = signatures;
+
+                if let Some(raw) = raw {
+                    mail_raw = raw;
+                    info!(context, "decrypted: {:?}", std::str::from_utf8(&mail_raw));
```

Comment:
can this not crash if the mail_raw is invalid utf8?

Reply:
it can, and it should not be there, I need to go and cleanup all my debugging logs

Reply:
are there many debugging logs? I think what's more important is going through all `from_utf8` if any.